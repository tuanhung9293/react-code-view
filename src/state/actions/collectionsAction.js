import {
  ADD_COLLECTION,
  EDIT_COLLECTION,
  REMOVE_COLLECTION,
  LIKE_COLLECTION,
  SORT_COLLECTION
} from 'constants/actionTypes'

export function addCollection(itemData) {
  return {
    type: ADD_COLLECTION,
    payload: {
      itemData: {...itemData, dateEdited: Date.now()}
    }
  }
}

export function editCollection(itemData) {
  return {
    type: EDIT_COLLECTION,
    payload: {
      itemData: {...itemData, dateEdited: Date.now()}
    }
  }
}

export function removeCollection(id) {
  return {
    type: REMOVE_COLLECTION,
    payload: {
      id
    }
  }
}

export function likeCollection(id) {
  return {
    type: LIKE_COLLECTION,
    payload: {
      id
    }
  }
}

export function sortNasaCollections(value) {
  const parsedValue = value.split(',')
  return {
    type: SORT_COLLECTION,
    payload: {
      ...parsedValue
    }
  }
}
