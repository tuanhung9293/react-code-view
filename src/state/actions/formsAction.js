import {
  SET_INITIAL_FORM_VALUES,
  SET_FORM_VALUE,
  FORM_FIELD_BLUR,
  VALIDATE_FORM,
  VALIDATE_FIELD,
  CLEAR_FORM
} from 'constants/actionTypes'

export function setInitialValues({form, initialValues}) {
  return {
    type: SET_INITIAL_FORM_VALUES,
    payload: {
      form,
      initialValues
    }
  }
}

export function setFormValue({form, name, value}) {
  return {
    type: SET_FORM_VALUE,
    payload: {
      form,
      name,
      value
    }
  }
}

export function formFieldBlur({form, name}) {
  return {
    type: FORM_FIELD_BLUR,
    payload: {
      form,
      name
    }
  }
}

export function validateForm({form, errors}) {
  return {
    type: VALIDATE_FORM,
    payload: {
      form,
      errors
    }
  }
}

export function validateField({form, name, error}) {
  return {
    type: VALIDATE_FIELD,
    payload: {
      form,
      name,
      error
    }
  }
}

export function clearForm({form}) {
  return {
    type: CLEAR_FORM,
    payload: {
      form
    }
  }
}