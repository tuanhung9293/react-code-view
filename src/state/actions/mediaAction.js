import axios from 'axios'
import {convertMedia, getThumbUrl, getMediaUrl} from 'services/medias'
import {AUDIO_THUMB_URL_DEFAULT} from 'constants/defaultValues'
import {SEARCH_API_URL} from 'constants/urls'
import {
  SEARCH_NASA_MEDIA,
  SEARCH_MORE_NASA_MEDIA,
  FETCH_MEDIA_ITEM_COLLECTION
} from 'constants/actionTypes'

export const searchNasaMedia = (params) => (dispatch) => {
  dispatch({
    type: SEARCH_NASA_MEDIA.PENDING,
    payload: {params}
  })
  return axios({
    method: 'get',
    url: `${SEARCH_API_URL}/search`,
    params: {
      ...params,
      year_start: '1920',
      year_end: '2018'
    }
  })
    .then((res) => {
      dispatch({
        type: SEARCH_NASA_MEDIA.SUCCESS,
        payload: {
          ...convertMedia(res.data),
          params: {...params, page: 1}
        }
      })
    })
}

export const searchMoreNasaMedia = (params) => (dispatch) => {
  dispatch({
    type: SEARCH_MORE_NASA_MEDIA.PENDING,
    payload: {params}
  })
  return axios({
    method: 'get',
    url: `${SEARCH_API_URL}/search`,
    params: {
      ...params,
      year_start: '1920',
      year_end: '2018'
    }
  })
    .then((res) => {
      dispatch({
        type: SEARCH_MORE_NASA_MEDIA.SUCCESS,
        payload: {...convertMedia(res.data), params: params}
      })
    })
}

export const fetchMediaItemCollection = (index, mediaType, collectionUrl) => (dispatch) => {
  dispatch({
    type: FETCH_MEDIA_ITEM_COLLECTION.PENDING,
    payload: {index, collectionUrl}
  })
  return axios({
    method: 'get',
    url: collectionUrl
  })
    .then((res) => dispatch({
      type: FETCH_MEDIA_ITEM_COLLECTION.SUCCESS,
      payload: {
        thumbUrl: mediaType === 'audio' ? AUDIO_THUMB_URL_DEFAULT : getThumbUrl(res.data),
        mediaUrl: getMediaUrl(res.data)
      },
      meta: {index: index}
    }))
}
