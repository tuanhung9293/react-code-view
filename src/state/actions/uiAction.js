
import {START_TRACKING_WINDOW, STOP_TRACKING_WINDOW, UPDATE_WINDOW_SIZE} from 'constants/actionTypes'

export function startTrackingWindow() {
  return {
    type: START_TRACKING_WINDOW
  }
}

export function stopTrackingWindow() {
  return {
    type: STOP_TRACKING_WINDOW
  }
}

export function updateWindowSize({width, height}) {
  return {
    type: UPDATE_WINDOW_SIZE,
    payload: {
      width,
      height
    }
  }
}