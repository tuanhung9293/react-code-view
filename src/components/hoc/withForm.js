import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {clearForm, validateForm, setInitialValues} from 'state/actions/formsAction'

export function createForm(formName, FormComponent) {
  class Form extends React.Component {
    validators = {}

    componentDidMount() {
      this.setInitialValues()
      this.validate()
    }

    componentWillUnmount() {
      this.clearForm()
    }

    shouldComponentUpdate(next) {
      const {submitting, submitted, valid, dirty} = this.props
      const shouldUpdate =
        submitting !== next.submitting ||
        submitted !== next.submitted ||
        valid !== next.valid ||
        dirty !== next.dirty
      return shouldUpdate
    }

    getChildContext() {
      return {
        _form: {
          name: formName,
          registerValidator: this.registerValidator,
          unregisterValidator: this.unregisterValidator
        }
      }
    }

    render() {
      const {
        form,
        ...rest
      } = this.props
      return <FormComponent form={form} formName={formName} {...rest} handleSubmit={this.handleSubmit} />
    }

    registerValidator = (name, validator) => {
      this.validators[name] = validator
    }

    unregisterValidator = (name) => {
      delete this.validators[name]
    }

    setInitialValues = () => {
      this.props.setInitialValues()
    }

    clearForm = () => {
      this.props.clearForm()
    }

    handleSubmit = () => {
      this.props.handleSubmit(this.props.form)
    }

    validate = () => {
      const {validateForm, form: {values}} = this.props
      const errors = Object.keys(this.validators).reduce((errors, name) => {
        const value = values[name]
        const validator = this.validators[name]
        const error = validator(value)
        if (error) {
          return {
            ...errors,
            [name]: error
          }
        }

        return errors
      }, {})
      validateForm(errors)
    }
  }

  Form.childContextTypes = {
    _form: PropTypes.object.isRequired
  }

  return Form
}

function withForm(
  formName = '',
  initialValues = {},
  submitActionCreator = (values = {}) => {},
  redirectAction = () => {},
  validator = (values = {}) => {}
) {
  function mapStateToProps({forms}) {
    const form = forms[formName] || {values: initialValues, touched: {}, validationErrors: {}, meta: {}}
    const {meta, ...rest} = form
    return {
      form: rest,
      ...meta,
      dirty: Object.keys(form.values).length > 0,
      valid: Object.keys(form.validationErrors).length === 0
    }
  }

  function mapDispatchToProps(dispatch) {
    return {
      setInitialValues: () => dispatch(setInitialValues({form: formName, initialValues})),
      clearForm: () => dispatch(clearForm({form: formName})),
      handleSubmit: (form) => {
        const values = {
          ...form.initialValues,
          ...form.values
        }
        dispatch(submitActionCreator(values))
        redirectAction()
      },
      validateForm: (errors) => dispatch(validateForm({form: formName, errors}))
    }
  }

  return (Component) => connect(mapStateToProps, mapDispatchToProps)(createForm(formName, Component))
}

export default withForm