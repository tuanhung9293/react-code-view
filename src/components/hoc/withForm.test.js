import React from 'react'
import withForm, {createForm} from './withForm'
import {shallow, mount} from 'enzyme'
import configureStore from 'redux-mock-store'
const mockStore = configureStore()
jest.mock('state/actions/formsAction')
import {setInitialValues, clearForm, validateForm} from 'state/actions/formsAction'

function createMockComponent() {
  return jest.fn().mockImplementation(props => null)
}

let MockComponent, store

beforeEach(() => {
  MockComponent = createMockComponent()
  store = mockStore({
    forms: {
      testForm: {
        values: {
          testField: 'changed value',
          another: 'value'
        },
        errors: {},
        validationErrors: {}
      }
    }
  })
  store.dispatch = () => {}
})

it('should take some configuration and return a component enhancer', () => {
  const enhance = withForm()
  const MockComponent = createMockComponent()
  const FormMockComponent = enhance(MockComponent)
  const enhancedComponent = shallow(<FormMockComponent store={store} />).dive()
  expect(enhancedComponent).toMatchSnapshot()
})

it('should dispatch set initial values action on mount', () => {
  const FormComponent = withForm('testForm', {testField: 'test value'})(createMockComponent())
  const wrapper = mount(<FormComponent store={store} />)
  expect(setInitialValues).toBeCalledWith({form: 'testForm', initialValues: {testField: 'test value'}})
})

it('should dispatch validate form action on mount', () => {
  const FormComponent = withForm('testForm', {testField: 'test value'})(createMockComponent())
  const wrapper = mount(<FormComponent store={store} />)
  expect(validateForm).toBeCalledWith({errors: {}, form: 'testForm'})
})

it('should dispatch clear form action on unmount', () => {
  const FormComponent = withForm('testForm', {testField: 'test value'})(createMockComponent())
  const wrapper = shallow(<FormComponent store={store} />).dive()
  wrapper.unmount()
  expect(clearForm).toBeCalled()
})

it('should dispatch submitActionCreator on submit', () => {
  const submitActionCreator = jest.fn()
  const FormComponent = withForm('testForm', {testField: 'test value'}, submitActionCreator)(createMockComponent())
  const wrapper = shallow(<FormComponent store={store} />).dive()
  wrapper.props().handleSubmit('this is the stuff')
  expect(submitActionCreator.mock.calls[0][0]).toEqual({
    testField: 'changed value',
    another: 'value'
  })
})

it('should append ids on submit', () => {
  const submitActionCreator = jest.fn()
  const FormComponent = withForm('testForm', {id: 123, testField: 'test value'}, submitActionCreator)(createMockComponent())
  const wrapper = shallow(<FormComponent store={store} />).dive()
  wrapper.props().handleSubmit('hello')
  expect(submitActionCreator.mock.calls[0][0]).toEqual({
    testField: 'changed value',
    another: 'value'
  })
})

it('should setup child context', () => {
  const FormComponent = createForm('testForm', createMockComponent())
  const wrapper = shallow(<FormComponent />, {disableLifecycleMethods: true})
  expect(wrapper.instance().getChildContext()).toMatchSnapshot()
})