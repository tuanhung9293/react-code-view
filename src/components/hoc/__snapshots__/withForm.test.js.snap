// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should setup child context 1`] = `
Object {
  "_form": Object {
    "name": "testForm",
    "registerValidator": [Function],
    "unregisterValidator": [Function],
  },
}
`;

exports[`should take some configuration and return a component enhancer 1`] = `
<mockConstructor
  clearForm={[Function]}
  dirty={false}
  form={
    Object {
      "touched": Object {},
      "validationErrors": Object {},
      "values": Object {},
    }
  }
  formName=""
  handleSubmit={[Function]}
  setInitialValues={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  valid={true}
  validateForm={[Function]}
/>
`;
